experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1066 @@
+                    return Error::invalid("Exceptions: missing closure data");
+                }
+
+                for c in self.closure_string(idx)?.chars() {
+                    next_idx += c.len_utf16() as u16;
+                }
+            }
+
+            valid_indices.push(idx);
+            idx = next_idx;
+        }
+        Ok(valid_indices)
+    }
+}
+
+#[test]
+fn test_exception_validation() {
+    let missing_slot_data: &[u16] = &[1 << (CaseMappingExceptionSlot::Lower as u16)];
+    let result = CaseMappingExceptions::try_from_icu(missing_slot_data);
+    assert_eq!(
+        result,
+        Err(Error::Validation("Exceptions: missing slot data"))
+    );
+
+    let invalid_char_slot: &[u16] = &[1 << (CaseMappingExceptionSlot::Lower as u16), 0xdf00];
+    let result = CaseMappingExceptions::try_from_icu(invalid_char_slot);
+    assert_eq!(
+        result,
+        Err(Error::Validation("Exceptions: invalid char slot"))
+    );
+
+    let missing_full_mappings: &[u16] =
+        &[1 << (CaseMappingExceptionSlot::FullMappings as u16), 0x1111];
+    let result = CaseMappingExceptions::try_from_icu(missing_full_mappings);
+    assert_eq!(
+        result,
+        Err(Error::Validation("Exceptions: missing full mapping data"))
+    );
+
+    let full_mapping_unpaired_surrogate: &[u16] = &[
+        1 << (CaseMappingExceptionSlot::FullMappings as u16),
+        0x1111,
+        0xdf00,
+    ];
+    let result = CaseMappingExceptions::try_from_icu(full_mapping_unpaired_surrogate);
+    assert!(matches!(result, Err(Error::DecodeUtf16(_))));
+
+    let missing_closure_data: &[u16] = &[1 << (CaseMappingExceptionSlot::Closure as u16), 0x1];
+    let result = CaseMappingExceptions::try_from_icu(missing_closure_data);
+    assert_eq!(
+        result,
+        Err(Error::Validation("Exceptions: missing closure data"))
+    );
+
+    let closure_unpaired_surrogate: &[u16] =
+        &[1 << (CaseMappingExceptionSlot::Closure as u16), 0x1, 0xdf00];
+    let result = CaseMappingExceptions::try_from_icu(closure_unpaired_surrogate);
+    assert!(matches!(result, Err(Error::DecodeUtf16(_))));
+}
+
+/// Reverse case folding data. Maps from multi-character strings back to code-points that fold to those
+/// strings.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, PartialEq, Clone, Yokeable, ZeroCopyFrom)]
+#[yoke(prove_covariance_manually)]
+struct CaseMappingUnfoldData<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    map: ZeroMap<'data, str, str>,
+}
+
+impl<'data> CaseMappingUnfoldData<'data> {
+    // Unfold data is exported by ICU as an array of 16-bit values, representing a short
+    // header followed by a two-column key/value table. The header indicates:
+    // - The number of rows.
+    // - The number of UTF16 code units per row.
+    // - The number of UTF16 code units in the first (key) column.
+    //   (The number of code units in the value column can be derived from the above.)
```

Comment:
Suggestion: I think this sort of decoding should be done in icuexportdata.  The TOML files we export should be "human readable" except for a small, specific set of exceptions, like CPTs.  If the data is really just a map from strings to strings, then the TOML should be a map of strings to strings.  Consumers of the TOML can readily massage the data into whatever format they like, including going back to an ICU4C-like structure.

If this is too much work, consider this suggestion optional.

Reply:
I'm going to skip this one. Given that the other two pieces of data in this file are a CPT (opaque) and a packed array of variable length data structures (extremely opaque), I don't think there's a lot of value in making this particular field human readable. I expect that the data provider on the ICU4X side will be the only consumer of this TOML.